src/contexts/Maci.tsx

```diff
@@ -0,0 +1,296 @@
+  const [tallyData, setTallyData] = useState<TallyData>();
+  const [nonce, setNonce] = useState<number>(0);
+
+  const attestations = api.voters.approvedAttestations.useQuery({
+    address,
+  });
+
+  const attestationId = useMemo(() => {
+    const values = attestations.data?.valueOf() as Attestation[] | undefined;
+    const length = values?.length ?? 1;
+
+    return values?.[length - 1]?.id;
+  }, [attestations]);
+
+  const isEligibleToVote = useMemo(
+    () => Boolean(attestationId) && Boolean(data),
+    [attestationId, attestations],
+  );
+
+  const votingEndsAt = useMemo(
+    () =>
+      pollData
+        ? new Date(
+            Number(pollData.deployTime) * 1000 +
+              Number(pollData.duration) * 1000,
+          )
+        : new Date(),
+    [pollData?.deployTime, pollData?.duration],
+  );
+
+  const onSignup = useCallback(
+    async (onError: () => void) => {
+      if (!data?.publicKey || !signer || !attestationId) {
+        return;
+      }
+
+      setIsLoading(true);
+
+      try {
+        const { stateIndex: index, hash } = await signup({
+          maciPubKey: data.publicKey,
+          maciAddress: config.maciAddress!,
+          sgDataArg: attestationId,
+          signer,
+        });
+
+        console.log(`Signup transaction hash: ${hash}`);
+
+        if (index) {
+          setIsRegistered(true);
+          setStateIndex(index);
+        }
+      } catch (e) {
+        onError();
+        console.error("error happened:", e);
+      } finally {
+        setIsLoading(false);
+      }
+    },
+    [
+      attestationId,
+      data?.publicKey,
+      address,
+      signer,
+      setIsRegistered,
+      setStateIndex,
+      setIsLoading,
+    ],
+  );
+
+  const onVote = useCallback(
+    async (
+      votes: IVoteArgs[],
+      onError: () => Promise<void>,
+      onSuccess: () => Promise<void>,
+    ) => {
+      if (!signer || !data || !stateIndex || !pollData) {
+        return;
+      }
+
+      if (!votes.length) {
+        await onError();
+        setError("No votes provided");
+        return;
+      }
+
+      const messages = votes.map(
+        ({ newVoteWeight, voteOptionIndex }, index) => ({
+          newVoteWeight,
+          voteOptionIndex,
+          stateIndex: BigInt(stateIndex),
+          maciContractAddress: config.maciAddress!,
+          nonce: BigInt(nonce + index + 1),
+        }),
+      );
+
+      setIsLoading(true);
+      await publishBatch({
+        messages,
+        maciContractAddress: config.maciAddress!,
+        publicKey: data.publicKey,
+        privateKey: data.privateKey,
+        pollId: BigInt(pollData.id),
+        signer: signer as unknown as Signer,
+      })
+        .then(() => onSuccess())
+        .catch((error: Error) => {
+          setError(error.message);
+          return onError();
+        })
+        .finally(() => {
+          setIsLoading(false);
+        });
+    },
+    [
+      nonce,
```

Comment:
nonce is not using in this file

Reply:
just thinking about maybe we need it to be set in the future, but could remove the useState of nonce currently